app/src/main/java/com/google/android/sambadocumentsprovider/ShareManager.java

```diff
@@ -64,56 +66,119 @@
+    final Map<String, ShareTuple> shareMap = new HashMap<>(serverStringSet.size());
     for (String serverString : serverStringSet) {
       // TODO: Add decryption
-      String uri = decode(serverString, credentialMap);
+      String uri = decode(serverString, shareMap);
       if (uri != null) {
         mServerStringMap.put(uri, serverString);
       }
     }
 
     mServerStringSet = new HashSet<>(serverStringSet);
 
-    for (Map.Entry<String, CredentialTuple> server : credentialMap.entrySet()) {
-      final CredentialTuple tuple = server.getValue();
+    for (Map.Entry<String, ShareTuple> server : shareMap.entrySet()) {
+      final ShareTuple tuple = server.getValue();
+
+      if (tuple.mIsMounted) {
+        mMountedServerSet.add(server.getKey());
+      }
+
       mCredentialCache.putCredential(
           server.getKey(), tuple.mWorkgroup, tuple.mUsername, tuple.mPassword);
     }
   }
 
-  public synchronized void mountServer(
-      String uri, String workgroup, String username, String password, ShareMountChecker checker)
-      throws IOException {
-    if (mServerStringMap.containsKey(uri)) {
-      throw new IllegalStateException("Uri " + uri + " is already mounted.");
+  /**
+   * Stores the server info (e.g. uri, credentials) permanently.
+   * @param mount the share will be displayed as a separate root.
+   * @param force if the share with such a uri already exists, it will be overwritten.
+   */
+  public synchronized void addServer(
+          String uri, String workgroup, String username, String password,
+          ShareMountChecker checker, boolean mount, boolean force) throws IOException {
+    if (!force && mServerStringMap.containsKey(uri)) {
+      if (mount) {
+        checkServerCredentials(uri, workgroup, username, password, checker);
+        setMount(uri, true);
+        return;
+      }
+
+      throw new IllegalStateException("Uri " + uri + " is already stored.");
     }
 
+    checkServerCredentials(uri, workgroup, username, password, checker);
+
     final boolean hasPassword = !TextUtils.isEmpty(username) && !TextUtils.isEmpty(password);
-    if (hasPassword) {
-      mCredentialCache.putCredential(uri, workgroup, username, password);
+    final ShareTuple tuple = hasPassword
+            ? new ShareTuple(workgroup, username, password, mount)
+            : ShareTuple.EMPTY_TUPLE;
+
+    updateServersData(uri, tuple, mount);
+  }
+
+  private void setMount(String uri, boolean mount)
+      throws IOException {
+    if (!mServerStringMap.containsKey(uri)) {
+      throw new IllegalStateException("Share with Uri " + uri + " not found.");
     }
-    try {
-      checker.checkShareMounting();
-    } catch (Exception e) {
-      Log.i(TAG, "Failed to mount server.", e);
-      mCredentialCache.removeCredential(uri);
-      throw e;
+
+    String serverString = mServerStringMap.get(uri);
+
+    Map<String, ShareTuple> shareMap = new HashMap<>(1);
+
+    decode(serverString, shareMap);
```

Comment:
Don't decode, just create a new `ShareTuple`.

Once loaded into memory, info in Java heap becomes the source of truth. Any potential discrepancy should be resolved to the info in Java heap.